text/0000-expand-math.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: Expand `math`
+- Start Date: 2022-02-28
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+This RFC proposes expanding the standard library's `math` package and the shape it might take to leverage the full extent of Pony. There was prior discussion on [Zulip](https://ponylang.zulipchat.com/#narrow/stream/192795-contribute-to.20Pony/topic/math.20lib) years ago, but I expect some opinions to have changed.
+
+# Motivation
+
+Currently, `math` includes limited functionality. This should be expanded to include math types, constants, and present a structure with further expansion in mind. An expanded `math` library will allow a unified mathematics among Pony developers.
+
+# Detailed design
+
+The primary goals of this initial expansion are: 
+
+1. restructure the `math` library into distinct sub-packages
+2. provide common `math` data types
+
+## Structure
+
+I propose a structure of distinct subpackages including the following:
+
++ `math/big`: Arbitrary precision numbers
++ `math/series`: Mathematical series
++ `math/constant`: Mathematical constants
++ `math/rational`: `Rational` data type and related functions
++ `math/complex`: `Complex` data type and related functions
++ `math/(x,exp,etc)`: experimental additions, utilities, and effective "catch-all" for matters that do not neatly fit into other subpackages
+
+## Common Data Types
+
+As previewed above in [Structure](#structure), expanding the `math` package should include implementations of common mathematics data types. Below are some implementation proposals for those data types.
+
+### `math/big`
+
+Should include `BigInt`, `BigFloat`, and `BigDecimal` -- see [168](https://github.com/ponylang/rfcs/issues/168)
+
+### `math/series`
+
+`math/series` should include a `Series` trait which is a subclass of `Iterator`. The purpose of creating a new abstract data type is to generalize functions over mathematical series which do not make sense over iterators -- such as whether a `Series` is diverging or converging, a property that is all but meaningless for an `Iterator`.
+
+Example series include `Fibonacci` (already exists), `Pascal` (nCk), `Triangular` ({n+1}C{2}), `Square` (n^2), `Pentagonal` ({2n * (2n - 1)} / 2), etc.
+
+Current `math/Fibonacci` would go into this series package.
+
+### `math/constant`
+
+Initial values to include are those with underlying LLVM representations from the [numbers namespace](https://llvm.org/doxygen/namespacellvm_1_1numbers.html).
+
+Once these values exist in `math/constant`, they should be removed from where they are now, which is on `F32` and `F64` of [`Float`](https://github.com/ponylang/ponyc/blob/master/packages/builtin/float.pony).
```

Comment:
Same thing I said to Joe:

> These do not necessarily need to move, but it is what I am suggesting. For me the reason is that we have previously had new Pony users (and experiences Pony users) be surprised by the availability of these constants on F32 and F64. I can expand the motivation to reflect this reasoning.

Reply:
Surprise doesn't seem like a good reason to me. I think what is most important would be usability. I'd like to hear how the move will make these constants more easily usable.

Reply:
It is a flimsy reason merely "it has surprised people before" I understand that.

What I would like to be able to do is have some guaranteed immutable value. As these are constants I would want to disallow mutations (ref and iso) entirely. I know generic constraints are possible, but have never explored a `#read` type return so not sure it is even possible. Is that possible?

Reply:
Best I can think of with existing knowledge is to "wrap" a Float and make all the mutating operations do nothing. But that seems like a bad idea even if it is possible.

```pony
class Constant [F: FloatingPoint[A] #read] is F
```

I highly doubt the above works, but that would be my idea of how to achieve this without compiler changes.

Reply:
So I think that `F32.pi()` is completely reasonable when you understand the language.

I want an F32 width of pi. 

What would this look like on Constant?

- `fun pi_f32()`
- `fun pi_f34()`

This is looking like a very large compiler change that I don't understand a good reason for.

So, all primitive values like F32 are already immutable. So I'm not sure what the "these are constant so I would want to disallow mutations". All F32's are already immutable.

Reply:
With the being on primitives they are already immutable so forget my comments on that.

As for what it would look like on Constant, I saw it as generic so for example `Constant[F64].pi() => F64(3.1415926...)`

Reply:
So if I do...

`Constant[U64].pi()` 

do I get 3?

This interface seems really weird to me. Pi is generally accepted "to be a float" (as far as computers are concerned", but there's an interface now to get it for me for non-floating ones.

Additionally, now, constants which will vary by underlying type are no longer on the underlying type but on this new primitive that groups things based on being a constant. So if we want to be able to get pi as a BigDecimal, we need to not only implement BigDecimal but we need to touch all of the code for constant to add all the appropriate constants. That feels wrong. Adding a new number type like BigDecimal should be relatively self-contained. This spreads it around. I'm not in favor of this without some serious convincing.

Reply:
I went against my own RFC here. The parameter is as the call site via `Constant.pi[A: Float](): A(3.14159...)`.

I changed "should" to "could" to address removal of methods on floats. The reason of "surprising placement" is not a good enough reason to remove them, but having a grouping of constants on a `Constant` primitive still seems reasonable to me.